src/Microsoft.ApplicationInsights.AspNetCore/DiagnosticListeners/Implementation/HostingDiagnosticListener.cs

```diff
@@ -41,10 +42,15 @@ public void OnBeginRequest(HttpContext httpContext, long timestamp)
                 {
                     Id = httpContext.TraceIdentifier
                 };
-
+                this.client.Initialize(requestTelemetry);
                 requestTelemetry.Start(/*timestamp*/);
-
                 httpContext.Features.Set(requestTelemetry);
+
+                IHeaderDictionary responseHeaders = httpContext.Response?.Headers;
+                if (responseHeaders != null && !responseHeaders.ContainsKey(RequestResponseHeaders.TargetInstrumentationKeyHeader) && !string.IsNullOrEmpty(requestTelemetry.Context.InstrumentationKey))
```

Comment:
Nit: For perf, put the ```IsNullOrEmpty``` check before the ```ContainsKey``` check. C# has guaranteed left-to-right evaluation and short-circuiting.